src/libraries/System.Transactions/src/System/Transactions/Oletx/DTCTransactionManager.cs

```diff
@@ -0,0 +1,181 @@
+            {
+                handle = HandleTable.AllocHandle(internalRM);
+
+                _proxyShimFactory.ConnectToProxy(
+                    _nodeName,
+                    internalRM.Identifier,
+                    handle,
+                    out nodeNameMatches,
+                    out _whereaboutsSize,
+                    out whereaboutsBuffer,
+                    out resourceManagerShim);
+
+                // If the node name does not match, throw.
+                if (!nodeNameMatches)
+                {
+                    throw new NotSupportedException(SR.ProxyCannotSupportMultipleNodeNames);
+                }
+
+                // Make a managed copy of the whereabouts.
+                if (whereaboutsBuffer != null && _whereaboutsSize != 0)
+                {
+                    _whereabouts = new byte[_whereaboutsSize];
+                    Marshal.Copy(whereaboutsBuffer.DangerousGetHandle(), _whereabouts, 0, Convert.ToInt32(_whereaboutsSize));
+                }
+
+                // Give the IResourceManagerShim to the internalRM and tell it to call ReenlistComplete.
+                internalRM.ResourceManagerShim = resourceManagerShim;
+                internalRM.CallReenlistComplete();
+
+                _initialized = true;
+            }
+            catch (COMException ex)
+            {
+                if (ex.ErrorCode == NativeMethods.XACT_E_NOTSUPPORTED)
+                {
+                    throw new NotSupportedException( SR.CannotSupportNodeNameSpecification);
+                }
+
+                OletxTransactionManager.ProxyException(ex);
+
+                // Unfortunately MSDTCPRX may return unknown error codes when attempting to connect to MSDTC
+                // that error should be propagated back as a TransactionManagerCommunicationException.
+                throw TransactionManagerCommunicationException.Create(SR.TransactionManagerCommunicationException, ex);
+            }
+            finally
+            {
+                if (whereaboutsBuffer != null)
+                {
+                    whereaboutsBuffer.Close();
+                }
+
+                // If we weren't successful at initializing ourself, clear things out
+                // for next time around.
+                if (!_initialized)
+                {
+                    if (handle != IntPtr.Zero && resourceManagerShim == null)
+                    {
+                        HandleTable.FreeHandle(handle);
+                    }
+
+                    if (null != _whereabouts)
+                    {
+                        _whereabouts = null;
+                        _whereaboutsSize = 0;
+                    }
+                }
+            }
+        }
+
+        internal IDtcProxyShimFactory ProxyShimFactory
+        {
+            get
+            {
+                if (!_initialized)
+                {
+                    lock (this)
+                    {
+                        Initialize();
```

Comment:
If Initialize is expensive you may want to recheck the `_initialized` value once the lock is taken.

Reply:
That check already happens inside Initialize.